### PR TITLE
virtualbox: fix audio driver loading

### LINKED ISSUE #123851: Virtualbox 6.1.22 fails to load libpulse.so
The pulseaudio backend stopped working with the update of Virtualbox 6.1.18 -> Virtualbox 6.1.22 (ba0da8a076e94ff1119cd46fe58d12c54f9e978b)

Virtualbox 6.1.20 [overhauled the audio device backend](https://www.virtualbox.org/wiki/Changelog), and changed the way libpulse.so is loaded 

**To Reproduce**
1. set up a VM with `pulseaudio` as a backend
2. start vm
3. you will get a popup about not being able to start audio backend

**Expected behavior**

Audio works! No Errors!

**Screenshots**

Here is the `VBox.log` for the session. 

```
    00:00:00.941279 HDA: Asynchronous I/O enabled
    00:00:00.941313 Audio: Initializing PulseAudio driver
    00:00:00.941320 audioLoadPulseLib: Failed to load library /nix/store/j8vl05zdjkp62pg23qrqra4dcahc77hg-libpulseaudio-14.2/lib/libpulse.so.0: VERR_INVALID_PARAMETER
    00:00:00.941344 PulseAudio: Failed to load the PulseAudio shared library! Error VERR_INVALID_PARAMETER
    00:00:00.941389 Audio: Initialization of host driver 'PulseAudio' failed with VERR_INVALID_PARAMETER
    00:00:00.941404 VM: Raising runtime error 'HostAudioNotResponding' (fFlags=0x0)
    00:00:00.941411 Console: VM runtime error: fatal=false, errorID=HostAudioNotResponding message="Host audio backend initialization has failed. Selecting the NULL audio backend with the consequence that no sound is audible"
    00:00:00.941509 HDA: Codec reset
    00:00:00.941533 HDA: Reset
```

@svanderburg 

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by jerith666**
fixes #123851

patch by @jcumming

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Note: I couldn't run the tests because the build for `VirtualBox-GuestAdditions-6.1.26-5.10.60` fails with:

```
  CC [M]  /build/install/src/vboxguest-6.1.26/vboxguest/common/math/gcc/qdivrem.o
In file included from /build/install/src/vboxguest-6.1.26/vboxguest/combined-os-specific.c:30:
/build/install/src/vboxguest-6.1.26/vboxguest/r0drv/linux/alloc-r0drv-linux.c: In function 'rtR0MemAllocEx':
/build/install/src/vboxguest-6.1.26/vboxguest/r0drv/linux/alloc-r0drv-linux.c:272:27: error: too many arguments to function '__vmalloc'
  272 |         pHdr = (PRTMEMHDR)__vmalloc(cb + sizeof(*pHdr), GFP_KERNEL | __GFP_HIGHMEM | __GFP_NOWARN, MY_PAGE_KERNEL_EXEC);
      |                           ^~~~~~~~~
In file included from /nix/store/iqv8g1m0xl18hc26cq0ik8kgqdfvzxxm-linux-5.10.60-dev/lib/modules/5.10.60/source/include/asm-generic/io.h:911,
                 from /nix/store/iqv8g1m0xl18hc26cq0ik8kgqdfvzxxm-linux-5.10.60-dev/lib/modules/5.10.60/source/arch/x86/include/asm/io.h:375,
                 from /nix/store/iqv8g1m0xl18hc26cq0ik8kgqdfvzxxm-linux-5.10.60-dev/lib/modules/5.10.60/source/include/linux/io.h:13,
                 from /nix/store/iqv8g1m0xl18hc26cq0ik8kgqdfvzxxm-linux-5.10.60-dev/lib/modules/5.10.60/source/include/linux/pci.h:39,
                 from /build/install/src/vboxguest-6.1.26/vboxguest/r0drv/linux/the-linux-kernel.h:135,
                 from /build/install/src/vboxguest-6.1.26/vboxguest/combined-os-specific.c:28:
/nix/store/iqv8g1m0xl18hc26cq0ik8kgqdfvzxxm-linux-5.10.60-dev/lib/modules/5.10.60/source/include/linux/vmalloc.h:107:14: note: declared here
  107 | extern void *__vmalloc(unsigned long size, gfp_t gfp_mask);
      |              ^~~~~~~~~
```

### CODE REVIEW

**Comment by jcumming** [comment]
Thanks for making a PR of this!

I tested this out using the pulseaudio backend, and it works. Have not checked the ALSA backend. 

I think this will close  #133341, #123851.

**Comment by Yarny0** [comment]
#### Review per https://nixos.org/manual/nixpkgs/stable/#chap-reviewing-contributions

* playing and recording audio works fine with this commit
* tested after rebasing the commit onto the current unstable channel commit (8a2ec31e224)
* tested with ALSA and with PulseAudio backend
* tested with NixOS 21.05 as VirtualBox guest

As a thought about https://github.com/NixOS/nixpkgs/issues/123851#issuecomment-845579963

> This may not be the *right* thing to do, but it works.

We're patching a couple of library paths in the `prePatch` phase, and VirtualBox developers might decide to change the load commands of more of those in the future.  My intuition says there might be a way to tell VirtualBox about a set of directories where it should search for its libraries.  Adding the paths of `alsa-lib` and `libpulseaudio` to that set instead of patching that full paths into the load commands should make `RTLdrLoadSystemEx` happy then.  However, I'm no expert in this regards -- maybe `patchelf` can help here, or some option for the compiler or linker.

Nevertheless, I recommend this commit to be merged so VirtualBox audio gets fixed.

**Comment by jerith666** [comment]
Also no expert, just codifying a known work-around.  I'm sure there's a better way to do this, and am happy to step out of the way if someone more knowledgeable proposes one.  :)

**Comment by parthy** [comment]
So could be removing those special path substitutions be the better solution? I have to verify properly, but it seems to work as well.

**Comment by otavio** [comment]
Could this to be considered for merging? this error is very annoying and we need to disable audio completely.

File: pkgs/applications/virtualization/virtualbox/default.nix [hunk 1]
```diff
@@ -97,6 +97,7 @@ in stdenv.mkDerivation {
     # Temporary workaround for broken build
     # https://www.virtualbox.org/pipermail/vbox-dev/2021-July/015670.html
     ./fix-configure-pkgconfig-qt.patch
+    ./fix-audio-driver-loading.patch
```
Comment (Artturin) [inline pkgs/applications/virtualization/virtualbox/default.nix:100]:
```suggestion
    # https://github.com/NixOS/nixpkgs/issues/123851
    ./fix-audio-driver-loading.patch
```

**Pull request merged**